apis/r/R/utils-arrow.R

```diff
@@ -348,21 +348,31 @@ extract_levels <- function(arrtbl, exclude_cols=c("soma_joinid")) {
 #' Domain and extent table creation helper for data.frame writes returning a Table with
 #' a column per dimension for the given (incoming) arrow schema of a Table
 #' @noRd
-get_domain_and_extent_dataframe <- function(tbl_schema, ind_col_names,
+get_domain_and_extent_dataframe <- function(tbl_schema, ind_col_names, domain = NULL,
                                             tdco = TileDBCreateOptions$new(PlatformConfig$new())) {
     stopifnot("First argument must be an arrow schema" = inherits(tbl_schema, "Schema"),
               "Second argument must be character" = is.character(ind_col_names),
               "Second argument cannot be empty vector" = length(ind_col_names) > 0,
               "Second argument index names must be columns in first argument" =
                   all(is.finite(match(ind_col_names, names(tbl_schema)))),
               "Third argument must be options wrapper" = inherits(tdco, "TileDBCreateOptions"))
+    if (!is.null(domain)) {
+      stopifnot("Non-null domain provided to create must have same length as index-column names" =
+        (length(domain) == length(ind_col_names)))
+      for (name in ind_col_names) {
+        slot <- domain[[name]]
+        if (!is.null(slot)) {
+          stopifnot("Non-null domain provided to create must have null or 2-element vector in each slot" =
+            (length(slot) == 2))
+        }
+      }
+    }
```

Comment:
Strengthend and simplified checking of `domain`
```suggestion
    stopifnot(
        "domain must be NULL or a named list, with values being 2-element vectors or NULL" = is.null(domain) ||
          ( # Check that `domain` is a list of length `length(index_column_names)`
            # where all values are named after `index_column_names`
            # and all values are `NULL` or a two-length atomic non-factor vector
            rlang::is_list(domain, n = length(index_column_names)) &&
              identical(sort(names(domain)), sort(index_column_names)) &&
              all(vapply_lgl(
                domain,
                function(x) is.null(x) || (is.atomic(x) && !is.factor(x) && length(x) == 2L)
              ))
          )
      )
```